### ISSUE TITLE
libnethogs total sent/recv counters missing closed connections

### ISSUE THREAD

**Issue by jantman** [raboof/nethogs#130]
@raboof first off, thank you so much for this great application.

I'm trying to implement graphing of per-process network traffic, similar to what #128 is talking about. I have a minimal example working using libnethogs via [contrib/python-wrapper.py](https://github.com/raboof/nethogs/blob/master/contrib/python-wrapper.py), as Python is my preferred language (and unfortunately, I can _barely_ understand the c++ source code). I modified that Python code a bit to change the output to a simpler one-line-per-callback format that I can grep (the current version of my code is [here](https://github.com/jantman/misc-scripts/blob/a67ab845fba05d8fa9c6b691c1406a7dc0530f3b/nethogs2statsd.py)).

As a test, I started my libnethogs (0.8.5) script and then ran a second Python script as my user that just gets 2048 bytes of data over HTTP every 40 seconds, three times:

```python
#!/usr/bin/env python

import os
import requests
from time import sleep

print('PID: %d' % os.getpid())

print('Getting 2048b over HTTP')
# read 2048b from httpbin.org
requests.get('https://httpbin.org/stream-bytes/2048').content
print('sleeping 40s')
sleep(40)
print('Getting 2048b over HTTP')
# read 2048b from httpbin.org
requests.get('https://httpbin.org/stream-bytes/2048').content
print('sleeping 20s')
sleep(40)
print('Getting 2048b over HTTP')
# read 2048b from httpbin.org
requests.get('https://httpbin.org/stream-bytes/2048').content
print('DONE.')
```

When I grep the output of my libnethogs wrapper for that PID, the results are somewhat confusing to me:

```
2017-08-24 05:58:52,053 record_id=7 action=SET name=python pid=32343 uid=1000 dev=enp4s0 sent_bytes=1280 recv_bytes=6108 sent_kbps=0.25 recv_kbps=1.192968726158142
2017-08-24 05:58:56,291 record_id=7 action=SET name=python pid=32343 uid=1000 dev=enp4s0 sent_bytes=1280 recv_bytes=6108 sent_kbps=0.0 recv_kbps=0.0
2017-08-24 05:59:32,049 record_id=7 action=SET name=python pid=32343 uid=1000 dev=enp4s0 sent_bytes=2494 recv_bytes=12150 sent_kbps=0.23710937798023224 recv_kbps=1.180078148841858
2017-08-24 05:59:36,079 record_id=7 action=SET name=python pid=32343 uid=1000 dev=enp4s0 sent_bytes=2494 recv_bytes=12150 sent_kbps=0.0 recv_kbps=0.0
2017-08-24 05:59:41,584 record_id=7 action=SET name=python pid=32343 uid=1000 dev=enp4s0 sent_bytes=1214 recv_bytes=6042 sent_kbps=0.0 recv_kbps=0.0
2017-08-24 06:00:12,071 record_id=7 action=SET name=python pid=32343 uid=1000 dev=enp4s0 sent_bytes=2428 recv_bytes=12085 sent_kbps=0.23710937798023224 recv_kbps=1.180273413658142
2017-08-24 06:00:16,077 record_id=7 action=SET name=python pid=32343 uid=1000 dev=enp4s0 sent_bytes=2428 recv_bytes=12085 sent_kbps=0.0 recv_kbps=0.0
2017-08-24 06:00:21,232 record_id=7 action=SET name=python pid=32343 uid=1000 dev=enp4s0 sent_bytes=1214 recv_bytes=6043 sent_kbps=0.0 recv_kbps=0.0
2017-08-24 06:01:01,049 record_id=7 action=SET name=python pid=32343 uid=1000 dev=enp4s0 sent_bytes=0 recv_bytes=0 sent_kbps=0.0 recv_kbps=0.0
2017-08-24 06:01:02,121 record_id=7 action=REMOVE name=python pid=32343 uid=1000 dev=enp4s0 sent_bytes=0 recv_bytes=0 sent_kbps=0.0 recv_kbps=0.0
```

It seems as though the ``sent_bytes`` and ``recv_bytes`` values keep going up and down. I've looked through the source code, but my c++ is so weak that I can't completely follow it. From the little bit of documentation that I found about the [NethogsMonitorRecord struct](https://github.com/raboof/nethogs/blob/0518642c032e97128cc625635c39e33bf84933c1/src/libnethogs.h#L21) that's passed to the callback, it seemed like ``sent_bytes`` and ``recv_bytes`` were supposed to be counters that always increase (unless their ``unit32`` wraps around)... so I'm not really sure why these values are going up and down?

Would it be possible for you to explain how that works?

My end goal is to find out how much data has transferred since the last callback, i.e. I want to execute code for each ``NethogsMonitorRecord.record_id`` that knows how much data has been sent and received since the last callback for that Record. Originally I thought that I could just subtract the last sent_bytes and recv_bytes for each record from the latest value, but that doesn't appear to work...

Any advice or help would be _greatly_ appreciated!

Thanks,
Jason

**Comment by jantman** [comment]
I've started trying to dig into the code for this, and with a lot of head scratching and recompiling with some printf statements, I think I may be figuring it out.

I very well may be wrong here, but it seems to me that with how Process and Connection work, the "total" that we come up with (i.e. that's passed to the libnethogs nethogsmonitor_loop callback) is only a total for _current_ connections; i.e. when a connection closes, it disappears from the Process's ConnList, and doesn't get added to the total anymore. If I'm right that would explain my behavior, but then it would also mean there's no actual way to come up with a real **total** for a process, and not even a way to tell how much data that process has transferred since the last SET action?

**Comment by jantman** [comment]
Yeah, that seems to be what's happening. I added in a bit of instrumentation that logs the length of each Process's ConnList. It appears that when a Connection is removed from the Process's ConnList, that data disappears from the total, i.e.:

```
2017-08-25 11:33:14,480 [DEBUG nethogs2statsd.py:182 - root._callback() ] record_id=6 action=SET name=b'/home/jantman/venvs/foo/bin/python' pid=15483 uid=1000 dev=enp4s0 sent_b=2560 recv_b=12283 sent_kbps=0.0 recv_kbps=0.0
Process::gettotal() PID=15483 current_connections_count=2 sum_sent=2560 sum_recv=12283
nethogsmonitor_handle_update() NEW TOTAL record_id=6 pid=15483 sent_bytes=2560 recv_bytes=12283
Process::gettotal() PID=15483 current_connections_count=2 sum_sent=2560 sum_recv=12283
nethogsmonitor_handle_update() NEW TOTAL record_id=6 pid=15483 sent_bytes=2560 recv_bytes=12283
Process::gettotal() PID=15483 current_connections_count=2 sum_sent=2560 sum_recv=12283
nethogsmonitor_handle_update() NEW TOTAL record_id=6 pid=15483 sent_bytes=2560 recv_bytes=12283
Process::gettotal() PID=15483 current_connections_count=2 sum_sent=2560 sum_recv=12283
nethogsmonitor_handle_update() NEW TOTAL record_id=6 pid=15483 sent_bytes=2560 recv_bytes=12283
Process::gettotal() PID=15483 current_connections_count=1 sum_sent=1280 sum_recv=6109
nethogsmonitor_handle_update() NEW TOTAL record_id=6 pid=15483 sent_bytes=1280 recv_bytes=6109
2017-08-25 11:33:19,174 [DEBUG nethogs2statsd.py:182 - root._callback() ] record_id=6 action=SET name=b'/home/jantman/venvs/foo/bin/python' pid=15483 uid=1000 dev=enp4s0 sent_b=1280 recv_b=6109 sent_kbps=0.0 recv_kbps=0.0
```

More specifically:

```
Process::gettotal() PID=15483 current_connections_count=2 sum_sent=2560 sum_recv=12283
nethogsmonitor_handle_update() NEW TOTAL record_id=6 pid=15483 sent_bytes=2560 recv_bytes=12283
Process::gettotal() PID=15483 current_connections_count=1 sum_sent=1280 sum_recv=6109
nethogsmonitor_handle_update() NEW TOTAL record_id=6 pid=15483 sent_bytes=1280 recv_bytes=6109
```

When a connection is removed from a Process's ConnList, its data is lost from the total.

I'm not really sure how to remedy this...

**Comment by raboof** [comment]
That sounds like a plausible observation, thanks for looking into this!

As for how to remedy this, I roughly see 2 ways forward:
* remove the 'cumulative bandwidth used' feature - nethogs is currently geared towards 'live' use anyway
* add fields to processes to capture the cumulative bandwidth used by since-cleaned-up connections

**Comment by jantman** [comment]
Ok, thanks! While my current use case isn't live, it's graphing, I've always found cumulative stats to be useful with nethogs and similar applications; I've had plenty of times when a strangely bursty application could end up hiding much of its traffic if the only stats are per-collection-interval.

On the other hand, looking at how the live totals are done, maybe I could just use that data and call it "good enough" for my use case.

I'll see if I can figure something out for this over the weekend. Thanks!

**Comment by jantman** [comment]
@raboof From what I can tell, this is a bug in libnethogs only, not nethogs itself, or else something that nethogs works around; I guess I'll have to dig in a bit deeper.

I followed your second suggestion, and [added fields](https://github.com/jantman/nethogs/commit/b471b3f371e1e551c58e980759857af2b1d04517) to Process to keep a running total of the sent/received bytes of cleaned-up connections, and then added fields to libnethogs' NethogsMonitorRecord for them. However, when I tested my code with both libnethogs and the nethogs executable, I saw something very interesting:

1. In one terminal, I started a python wrapper around libnethogs that just logs the update records for all activity on my primary Ethernet interface (enp4s0). In another terminal, I ran ``nethogs -v 2 enp4s0``.
2. I let both run for a few seconds to stabilize, then launched a new browser (Firefox) with an empty session, so that both nethogs and libnethogs would see the new process and connections at around the same time. Then I browsed to YouTube and started watching a random video. After about 60 seconds, I browsed to my blog (simple, hosted on GitHub Pages, no ads or media).
3. Within about a minute, the connections to youtube & related hosts started getting cleaned up. What I saw, though (screenshot below), was really unexected: ``nethogs -v 2`` still showed the correct overall total of sent and received bytes (1118223b and 26456448b, respectively), but for the same process libnethogs was reporting sent_b=59896 and recv_b=96437. With my patch to report on closed connections, however, it was also reporting sent_by_closed_bytes=1056788 recv_by_closed_bytes=26358553. The sum of these values is 1116684 and 26454990, much closer to what ``nethogs -v 2`` reports (I'd call it identical, since the two processes were started at slightly different times, and therefore their polling intervals are offset).

__From what I can tell given the above, it looks like ``nethogs -v 2`` counts closed connections towards the total, but ``libnethogs`` ignores them.__

**Comment by jantman** [comment]
I guess I should apologize for such a long issue, but at this point I'm relatively obsessed with this, and every time I think I'm done I dig into it a bit more.

__I've finally gotten to the root of this:__

* [Process::getkbps](https://github.com/raboof/nethogs/blob/99a38c003c6317a910ba8ea747be770940acdebd/src/process.cpp#L96) is what handles deleting closed/stalled connections from the ConnList. Process::gettotal just iterates over the ConnList (without modifying it or checking for stale/closed connections) and sums up the sent and received values for each connection.
* The code path run by the ``nethogs`` executable, in the ``do_refresh()`` loop in cui.cpp, [only calls getkbps() in VIEWMODE_KBPS](https://github.com/raboof/nethogs/blob/99a38c003c6317a910ba8ea747be770940acdebd/src/cui.cpp#L403); if it's running in one of the VIEWMODE_TOTAL modes, ``Process::gettotal`` is called without ever calling ``Process::getkbps``; this results in closed/stale connections never being deleted from ConnList, so the "total" counts that are displayed in the UI are always the actual total for all connections.
* In libnethogs, however, [both Process::getkbps and Process::gettotal are called for every update](https://github.com/raboof/nethogs/blob/99a38c003c6317a910ba8ea747be770940acdebd/src/libnethogs.cpp#L211); the result of this is that, unlike the ``nethogs`` executable, getkbps() cleans up the ConnList and then gettotal() only sees open connections.

This can be confirmed relatively trivially with the following patch to current master that stops libnethogs from calling Process::getkbps:

```diff
diff --git a/src/libnethogs.cpp b/src/libnethogs.cpp
index 18a30e2..0331a9e 100644
--- a/src/libnethogs.cpp
+++ b/src/libnethogs.cpp
@@ -206,9 +206,9 @@ static void nethogsmonitor_handle_update(NethogsMonitorCallback cb) {
       const u_int32_t uid = curproc->getVal()->getUid();
       u_int32_t sent_bytes;
       u_int32_t recv_bytes;
-      float sent_kbs;
-      float recv_kbs;
-      curproc->getVal()->getkbps(&recv_kbs, &sent_kbs);
+      float sent_kbs = 0.0;
+      float recv_kbs = 0.0;
+      // curproc->getVal()->getkbps(&recv_kbs, &sent_kbs);
       curproc->getVal()->gettotal(&recv_bytes, &sent_bytes);
 
       // notify update
```

When I repeat my browser test with that patch applied, libnethogs always reports sent and received kbps values of 0, but the total values almost exactly match what the ``nethogs`` UI shows.

I can think of three possible solutions to this:

1. The one that I've implemented in my [total-count-debug branch](https://github.com/raboof/nethogs/compare/master...jantman:total-count-debug) that adds ``sent_by_closed_bytes`` and ``recv_by_closed_bytes`` fields to Process, sets them in ``Process::getkbps``, then exposes them via a new ``Process::gettotalbyclosedconns()`` which is called in ``libnethogs nethogsmonitor_handle_update()``, which in turn sets similar new fields in ``NethogsMonitorRecord`` to those values. This results in a change to the ``NethogsMonitorRecord`` struct.
2. Retain those counters in Process, but have libnethogs set ``NethogsMonitorRecord.sent_bytes`` and ``.recv_bytes`` to the _sum_ of the respective results from ``Process::gettotal()`` and the new ``Process::gettotalbyclosedconns()``. This would preserve the public API of libnethogs NethogsMonitorRecord, but have the ``sent_bytes`` and ``recv_bytes`` fields now include closed connections as well, like the nethogs UI in total mode.
3. Make libnethogs work like the UI where it operates in _either_ kbps or total mode and calls _either_ ``Process::getkbps()`` _or_ ``Process::gettotal()`` depending on what mode it's operating in.

**Comment by raboof** [comment]
Ha, good observation, thanks!

With the `sent_by_closed_bytes` and `rcvd_by_closed_bytes` fields you added, wouldn't it be possible to make the `gettotal` call(s) return the correct amount without introducing `gettotalbyclosedconns`?

**Comment by jantman** [comment]
@raboof yeah! Sorry, I was working through a bunch of this stuff and didn't see your comment until now.

Yeah, I could do that as well as long as I initialized them to 0 properly. It would mean that the change also gets executed by ``nethogs`` not just libnethogs, but for ``nethogs`` they'd always be zero.

**Comment by raboof** [comment]
Fixed by #133 - great contribution, thanks a lot!

**Comment by jantman** [comment]
Sure, thanks so much for merging it! I'm actually working on #119 right now, but I think that's going to be my last PR for the time being.
